Only return iOS voices that work

talkr-app opened the pull request:
Only one voice per locale works in iOS (versions 12 & 13 and maybe
earlier versions ). This change brings back the ios.js file and some of
its former code to detect iOS.  It deletes the iOS8 & iOS9 voices
from ios.js in the lib folder (was that erroneously left over?) It then
filters the voices returned by iOS and only returns the ones that will
work.

talkr-app committed:
Only return iOS voices that work

Only one voice per locale works in iOS (versions 12 & 13 and maybe
earlier versions ). This change brings back the ios.js file and some of
its former code to detect iOS.  It deletes the iOS8 & iOS9 voices
from ios.js in the lib folder (was that erroneously left over?) It then
filters the voices returned by iOS and only returns the ones that will
work.

tom-s commented on src/ios.js:
```diff
@@ -0,0 +1,55 @@
+export const iOSversion = () => {
```
it seems we don't need the version ? What do you think about something like that ?
```
isIos = () => /(iPhone|iPad|iPod)/.test(navigator.platform)
```

talkr-app replied:
Yeah, works for me.  Hopfully Apple will fix it in iOS 14 and we'll have to change it back, but I think that's optimistic.  This has been in there a while.

tom-s commented on src/ios.js:
```diff
@@ -0,0 +1,55 @@
+    'Sin-Ji',
+    'Mei-Jia'
+  ]
+  return voices.filter((v) => {
```
```
return voices.filter(v => selectableVoices.includes(v.name))
```

tom-s reviewed (commented):
I've suggested a few changes, thanks a lot for  your PR

talkr-app committed:
Add splitPhrases option

talkr-app committed:
speak resolves/rejects with null onend & onerror

talkr-app committed:
compile previous change

tom-s commented on lib/speak-tts.js:
```diff
@@ -36,6 +36,7 @@ function () {
 
         var listeners = (0, _utils.isNil)(conf.listeners) ? {} : conf.listeners;
         var splitSentences = (0, _utils.isNil)(conf.splitSentences) ? true : conf.splitSentences;
+        var splitPhrases = (0, _utils.isNil)(conf.splitPhrases) ? true : conf.splitPhrases;
```
what is this option for ? How does it compare to splitSentences ?

talkr-app replied:
Sorry, looks like I messed up working with this PR and included my internal changes.  Not sure how useful this will be to the general public, but I'm trying to sync a moving picture to the TTS and I need to know when there is talking and when there is silence.  splitPhrases splits on other punctuation characters that cause a pause in the TTS (like the comma).

tom-s replied:
alright, i'd be grateful if you could remove it. Even better, splitSentences could probably be refactored so that it can serve both purposes (by passing a list of splitting chars for instance) ?

talkr-app replied:
A custom list of splitting chars would work for me and keep me from having to use a separate branch.  I'll work on that.  

Happy to remove this change, but I'm just not too familiar working with PRs...that's how I ended up with this in here in the first place :)

tom-s replied:
I think I 'll do a massive refactor of this lib in the next month anyway, I'll include your changes. In the meantime you can use your own version and close this PR, or update the PR and I'll merge it.
It's up to you, let me know !

talkr-app replied:
ok, removed the splitPhrases option.  I'll work on making splitSentences take a custom list of splitting chars

tom-s commented on src/ios.js:
```diff
@@ -0,0 +1,55 @@
+    'Mei-Jia'
+  ]
+  return voices.filter((v) => {
+    if(selectableVoices.includes(v.name)) {
```
You can make it a bit simpler
````
 return voices.filter((v) => selectableVoices.includes(v.name))
````

tom-s commented on src/speak-tts.js:
```diff
@@ -166,8 +188,8 @@ class SpeakTTS {
 
-        // Attach event listeners
-        Object.keys(listeners).forEach(listener => {
+        // Attach event listeners, and make sure onerror and onend are included
```
How is that necessary  ?

talkr-app replied:
onerror & onend are used by speaktts to resolve/reject the speak promise, and this should happen even if the user did not pass customized versions of those liseteners into the init function.

tom-s replied:
if no customized listener is passed, this is going to attach undefined variables ?

talkr-app replied:
No, newListener is a function that calls the user-provided listener (in this scenerio that is NULL and it won't be called due to "fn && fn(data)" ) but it also checks if the listener is for "onend" or "onerror" in order to resolve/reject the speak promise.

Prior to this change, if no custom listener for onend and onerror is passed, then the speak promise will not resolve or reject.

tom-s replied:
I see, thanks for the explanation. I think it could be resolved by giving a noop function as a defaut value for onEnd et onError

tom-s commented on src/ios.js:
```diff
@@ -0,0 +1,55 @@
+  return false
+}
+export const filterIOSVoices = (voices) => {
+  let selectableVoices = [
```
const

tom-s commented on src/utils.js:
```diff
@@ -5,6 +5,16 @@ export const splitSentences = (text = '') => text.replace(/\.+/g,'.|')
   .map(sentence => trim(sentence))
   .filter(Boolean)
 
+export const splitPhrases = (text = '') => text.replace(/\.+/g,'.|')
```
this can be deleted I think

talkr-app committed:
Remove splitPhrases change

talkr-app committed:
Update Readme & remove splitPhrases from utils

talkr-app committed:
Allow custom splitSentences characters

splitSentences can now be true, false, or set to a string of custom
characters to deliminate the sentences.

An error is thrown if [a-zA-Z] are passed (a user sending in 'true'
could be in for a world of hurt  figuring that out.)

Any number of consecutive splitting characters can be used, and
the sentence will only be split once.
